### PR TITLE
add nv-codec-headers

### PR DESCRIPTION
**Pull request by dvirtz**
Specify library name and version:  **nv-codec-headers/9.1.23.2**

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/nv-codev-headers/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        return self._autotools
+
+    def build(self):
+        autotools = self._configure_autotools()
+        autotools.make()
+
+    def package(self):
+        autotools = self._configure_autotools()
+        vars = autotools.vars
+        autotools.install(args=["PREFIX={}".format(self.package_folder)])
```
Comment (uilianries) [inline recipes/nv-codev-headers/all/conanfile.py:37]:
```suggestion
        autotools.install(args=["PREFIX={}".format(self.package_folder)])        
```

Need to extract the license from header: https://docs.conan.io/en/latest/howtos/extract_licenses_from_headers.html

File: recipes/nv-codev-headers/all/test_package/test_package.cpp [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/* example was taken from https://www.gnu.org/ghm/2011/paris/slides/andreas-enge-mpc.pdf */
+
+#include <ffnvcodec/nvEncodeAPI.h>
+#include <iostream>
+
+int main () {
+    std::cout << "hello NVENC API version " << NVENCAPI_MAJOR_VERSION << "." << NVENCAPI_MINOR_VERSION << "\n";
```
Comment (madebr) [inline recipes/nv-codev-headers/all/test_package/test_package.cpp:7]: This only tests the header: `NVENCAPI_MAJOR_VERSION` and NVENCAPI_MAJOR_VERSION` are macros.
Please call some function such that the actual library is tested.

File: recipes/nv-codev-headers/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    def _configure_autotools(self):
+        if self._autotools:
+            return self._autotools
+        self._autotools = AutoToolsBuildEnvironment(self)
+        return self._autotools
+
+    def build(self):
+        autotools = self._configure_autotools()
+        autotools.make()
```
Comment (madebr) [inline recipes/nv-codev-headers/all/conanfile.py:32]: I think you need to add some build requirements to get it building on Windows.
`make` is not available by default on Windows.

File: recipes/nv-codec-headers/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    def package_id(self):
+        self.info.header_only()
+
+    def build_requirements(self):
+        if tools.os_info.is_windows:
+            if "CONAN_MAKE_PROGRAM" not in os.environ and not tools.which("make"):
+                self.build_requires("make/4.2.1")
+
+    def source(self):
```
Comment (madebr) [inline recipes/nv-codec-headers/all/conanfile.py:26]: This has the possibility to break on c3i.
Also, I am not sure about your test: `if "CONAN_MAKE_PROGRAM" not in os.environ and not tools.which("make")`.
`make` on Windows does provide `gnumake.exe`, not `make.exe`.


```suggestion
    _autotools = None

    @property
    def _settings_build(self):
        return self.settings_build if hasattr(self, "settings_build") else self.settings

    def build_requirements(self):
        if self._settings_build.os == "Windows":
            if not tools.get_env("CONAN_MAKE_PROGRAM") and not tools.which("make"):
                self.build_requires("make/4.2.1")

    def package_id(self):
        self.info.header_only()

    def source(self):
```

File: recipes/nv-codec-headers/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        tools.save(os.path.join(self.package_folder, "licenses", "LICENSE"), license_contents)
+
+        # Package it
+        self.copy("license*", dst="licenses",  ignore_case=True, keep_path=False)
+
+        autotools = self._configure_autotools()
+        vars = autotools.vars
+        autotools.install(args=["PREFIX={}".format(self.package_folder)])
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (madebr) [inline recipes/nv-codec-headers/all/conanfile.py:52]: Please add a `package_info` method that models the `.pc` file.
A `ffnvcodec.pc` file is installed.

File: recipes/nv-codec-headers/all/test_package/test_package.cpp [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/* example was taken from https://www.gnu.org/ghm/2011/paris/slides/andreas-enge-mpc.pdf */
+
+#include <ffnvcodec/nvEncodeAPI.h>
+#include <iostream>
+
+int main () {
+    std::cout << "hello NVENC API version " << NVENCAPI_MAJOR_VERSION << "." << NVENCAPI_MINOR_VERSION << "\n";
+}
```
Comment (madebr) [inline recipes/nv-codec-headers/all/test_package/test_package.cpp:8]: We tend to use c examples as much as possible.
Please rename this file to `test_package.c` + modify `test_package/CMakeLists.txt`.
This tests whether nv-codec-header is really a c header and isn't dependent on any c++ standard library.

```suggestion
/* example was taken from https://www.gnu.org/ghm/2011/paris/slides/andreas-enge-mpc.pdf */

#include "ffnvcodec/nvEncodeAPI.h"
#include <stdio.h>

int main () {
    printf("hello NVENC API version %d.%d\n", NVENCAPI_MAJOR_VERSION, NVENCAPI_MINOR_VERSION);
}
```

File: recipes/nv-codec-headers/all/conanfile.py (hunk 5 above)
Comment (madebr) [reply recipes/nv-codec-headers/all/conanfile.py:47]: Configure scripts often use pc files to check whether a dependency is available.
So this is kinda important.

Comment (dvirtz) [reply recipes/nv-codec-headers/all/conanfile.py:47]: I don't understand what you're suggesting. Aren't pc files [forbidden](https://github.com/conan-io/hooks/blob/5ac6c12edad4f258dab4fbb37cd4be3ea5bffa58/hooks/conan-center.py#L841) in Conan center?

Comment (madebr) [reply recipes/nv-codec-headers/all/conanfile.py:47]: They aren't forbidden.
But we only explicitly set the `pkg_config` name when the upstream project creates one.

File: recipes/nv-codec-headers/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    def package_id(self):
+        self.info.header_only()
+
+    def build_requirements(self):
+        if tools.os_info.is_windows:
+            if "CONAN_MAKE_PROGRAM" not in os.environ:
+                self.build_requires("make/4.2.1")
+
+    def source(self):
```
Comment (madebr) [inline recipes/nv-codec-headers/all/conanfile.py:24]: Use `self.settings_build` (if available) for cross building support.
```suggestion
    _autotools = None

    @property
    def _settings_build(self):
        return getattr(self, "settings_build", self.settings)

    def build_requirements(self):
        if self._settings_build.os == "Windows":
            if "CONAN_MAKE_PROGRAM" not in os.environ:
                self.build_requires("make/4.2.1")

    def package_id(self):
        self.info.header_only()

    def source(self):
```

File: recipes/nv-codec-headers/all/test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
```
Comment (madebr) [inline recipes/nv-codec-headers/all/test_package/CMakeLists.txt:2]:
```suggestion
project(test_package C)
```

File: recipes/nv-codec-headers/all/conanfile.py (hunk 7 above)
Comment (dvirtz) [reply recipes/nv-codec-headers/all/conanfile.py:24]: this is a header only library, it has no settings

Comment (madebr) [reply recipes/nv-codec-headers/all/conanfile.py:24]: header-only libraries can have a settings object.
As long as you clear it in `package_id`.

File: recipes/nv-codec-headers/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    def build_requirements(self):
+        if self._settings_build.os == "Windows":
+            if "CONAN_MAKE_PROGRAM" not in os.environ:
+                self.build_requires("make/4.2.1")
+
+    def package_id(self):
+        self.info.header_only()
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version], strip_root=True)
```
Comment (madebr) [inline recipes/nv-codec-headers/all/conanfile.py:30]: We usually extract the sources in a `self._source_subfolder` subdirectory, but I guess it's fine.

**Review by madebr — APPROVED** [review]

File: recipes/nv-codec-headers/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+            autotools.make()
+
+    def package(self):
+        # Extract the License/s from the header to a file
+        tmp = tools.load(os.path.join(self._source_subfolder, "include", "ffnvcodec", "nvEncodeAPI.h"))
+        license_contents = tmp[2:tmp.find("*/", 1)] # The license begins with a C comment /* and ends with */
+        tools.save(os.path.join(self.package_folder, "licenses", "LICENSE"), license_contents)
+
+        autotools = self._configure_autotools()
+        vars = autotools.vars
```
Comment (uilianries) [inline recipes/nv-codec-headers/all/conanfile.py:51]:
```suggestion
```

Unused

File: recipes/nv-codec-headers/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        autotools = self._configure_autotools()
+        with tools.chdir(os.path.join(self.build_folder, self._source_subfolder)):
+            autotools.make()
+
+    def package(self):
+        # Extract the License/s from the header to a file
+        tmp = tools.load(os.path.join(self._source_subfolder, "include", "ffnvcodec", "nvEncodeAPI.h"))
+        license_contents = tmp[2:tmp.find("*/", 1)] # The license begins with a C comment /* and ends with */
+        tools.save(os.path.join(self.package_folder, "licenses", "LICENSE"), license_contents)
+
```
Comment (uilianries) [inline recipes/nv-codec-headers/all/conanfile.py:49]:
```suggestion
    def _extract_license(self):
        tmp = tools.load(os.path.join(self._source_subfolder, "include", "ffnvcodec", "nvEncodeAPI.h"))
        license_contents = tmp[2:tmp.find("*/", 1)]
        tools.save(os.path.join(self.package_folder, "licenses", "LICENSE"), license_contents)

    def package(self):
        self._extract_license()
```

Keep license extraction in a separated method.

**Review by uilianries — COMMENTED** [review]
It looks good, only small details.

**Review by uilianries — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**